### PR TITLE
Finished algorithm for meetings query

### PR DESCRIPTION
**Pull request by shaargtz**
Removed edge case logic since the algorithm by itself takes care of them

### CODE REVIEW

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 1]
```diff
@@ -14,10 +14,67 @@
...
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
 
 public final class FindMeetingQuery {
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:24]: Can you add class-level comments?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 2]
```diff
@@ -14,10 +14,67 @@
...
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:25]: Can you add comments for this method?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 3]
```diff
@@ -14,10 +14,67 @@
...
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<Event> eventList = new ArrayList(events);
+
+    Collections.sort(eventList, new Comparator<Event>() {
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:28]: Can you clarify in a comment that you're sorting by event start time?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 4]
```diff
@@ -14,10 +14,67 @@
...
+      @Override
+      public int compare(Event event1, Event event2) {
+        return event1.getWhen().start() - event2.getWhen().start();
+      }
+    });
+
+    int currentRangeStart = TimeRange.START_OF_DAY;
+    int currentRangeEnd;
+    int currentRangeDuration;
+
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:38]: No need for this space. I'm going to mark this in other places too, so my apologies for the copy & paste :)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 5]
```diff
@@ -14,10 +14,67 @@
...
+    boolean sharesAttendees;
+
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:40]: No need for this space.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 6]
```diff
@@ -14,10 +14,67 @@
...
+    List<TimeRange> availableTimes = new ArrayList<>();
+
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:42]: No need for this space.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 7]
```diff
@@ -14,10 +14,67 @@
...
+    for (Event event : eventList) {
+      sharesAttendees = false;
+
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:45]: No need for this space.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 8]
```diff
@@ -14,10 +14,67 @@
...
+    int currentRangeStart = TimeRange.START_OF_DAY;
+    int currentRangeEnd;
+    int currentRangeDuration;
+
+    boolean sharesAttendees;
+
+    List<TimeRange> availableTimes = new ArrayList<>();
+
+    for (Event event : eventList) {
+      sharesAttendees = false;
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:44]: You should declare this variable here, rather than outside the for loop, since you never reference it outside the for loop.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 9]
```diff
@@ -14,10 +14,67 @@
...
+
+      for (String person : event.getAttendees()) {
+        if (request.getAttendees().contains(person)) {
+          sharesAttendees = true;
+          break;
+        }
+      }
+
+      if (!sharesAttendees) continue;
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:53]: Your logic is a little confusing to follow for me. Consider adding helper functions to these loops - for instance this would look a lot more readable like the following:

```
for (Event event : eventList) {
  if (!eventHasRequestedAttendee(event, event.getAttendees())) {
    // This event does not conflict with any requested attendees' schedules.
    continue;
  }  
}
```

Consider how you can utilize this here and elsewhere in your function to increase readability.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 10]
```diff
@@ -14,10 +14,67 @@
...
+
+      currentRangeEnd = event.getWhen().start();
+      currentRangeDuration = currentRangeEnd - currentRangeStart;
+
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:57]: No need for a space here.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 11]
```diff
@@ -14,10 +14,67 @@
...
+      if (currentRangeEnd >= currentRangeStart) {
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:58]: When is this possible? Can you comment on that?

Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:58]: When is this possible? Can you comment on this?

Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:58]: Is it worth checking equality here? Will there ever be a duration request of '0'? If this is to satisfy that edge case, perhaps it's worth special-casing it so it's more apparent to someone reading the algorithm that this unexpected input is handled.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 12]
```diff
@@ -14,10 +14,67 @@
...
+        if (currentRangeDuration >= request.getDuration()) {
+          availableTimes.add(
+            TimeRange.fromStartDuration(currentRangeStart, currentRangeDuration));
+        }
+        currentRangeStart = event.getWhen().end();
+      } else {
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:64]: When is this possible? Can you comment on this? I think it's when events overlap, but the names of your variables make it sound like this is impossible. Perhaps consider different names for these variables to avoid this confusion.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 13]
```diff
@@ -14,10 +14,67 @@
...
+        if (event.getWhen().end() > currentRangeStart) {
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:65]: You can combine this into an 'else if' statement!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 14]
```diff
@@ -14,10 +14,67 @@
...
+          currentRangeStart = event.getWhen().end();
+        }
+      }
+    }
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:69]: The whole above section could use some clarifying comments or helper functions. This was pretty confusing to me - I figured it out and it's actually pretty nifty, but it's hard for another person to parse at the moment.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 15]
```diff
@@ -14,10 +14,67 @@
...
+
+    currentRangeEnd = TimeRange.END_OF_DAY;
+    currentRangeDuration = currentRangeEnd - currentRangeStart;
+    if (currentRangeDuration >= request.getDuration()) {
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:73]: Maybe just skip the currentRangeDuration assignment above and just check 'if (currentRangeEnd - currentRangeStart >= request.getDuration())'?

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 16]
```diff
@@ -14,10 +14,67 @@
...
+      availableTimes.add(
+        TimeRange.fromStartEnd(currentRangeStart, currentRangeEnd, true));
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:75]: nit: this isn't in the style guide but it's good practice for functions that use string or boolean parameters, and it's not immediately clear what that parameter is doing at call time, to annotate it, like `TimeRange.fromStartEnd(currentRangeStart, currentRangeEnd, /*inclusive=*/true)`

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 17]
```diff
@@ -14,10 +14,67 @@
...
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<Event> eventList = new ArrayList(events);
+
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:27]: You can get rid of this empty line.

**Review by timothy-parsons — CHANGES_REQUESTED** [review]
Hey Shaar! This is a very cool algorithm! Make sure it's well-documented and well-clarified, because at times it was hard to parse and follow. Good work!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 1 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:28]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 2 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:39]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 3 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:53]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 4 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 5 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 6 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 7 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 8 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 10 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 16 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 15 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:107]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 11 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:82]: It is not so much about a request duration of zero, but rather the case when event B starts right when event A ends, meaning that there is no space between them, but it indeed is more readable to have it as a separate case.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 12 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: You're right, I was thinking of the available time as a range with a beginning and end, but it is a bit more confusing and not very descriptive to think about it this way.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 13 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 9 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 14 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:106]: Sorry, I forgot to document because I was trying to get it done quickly, I forgot that's not the way to do things. I think comments, variable names and structure makes everything a lot clearer now, but I don't know which parts can I move to a helper function and actually make code more readable without extremely long lines. We can talk a bit more about this in tomorrow's 1:1.

**Comment by shaargtz** [comment]
Yeah, I forgot to optimize more for the reader ahaha sorry about that, hope it's ok now.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 18]
```diff
@@ -14,10 +14,118 @@
...
+   * availableTimeDuration is set to the available time from the end of the
+   * previous event until the start of the current event. Then, if
+   * availableTimeDuration is at least the same size as the duration of the
+   * request, a TimeRange from pastEventEnd to currentEventStart is
+   * added to the list of available times.
+   *
+   * Time complexity: 
+   * n: number of events
+   * m: number of attendees in every event
+   * If m > logn, then O(nm)
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:50]: Always assume worst case for big-O analysis. So, this would simply be O(nm)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 19]
```diff
@@ -14,10 +14,118 @@
...
+  };
+  /**
+   * The way the query works is by going through the sorted events, and
+   * if the current event does not have any of the people in the meeting
+   * request, the event is skipped. If it does share people with the request,
+   * availableTimeDuration is set to the available time from the end of the
+   * previous event until the start of the current event. Then, if
+   * availableTimeDuration is at least the same size as the duration of the
+   * request, a TimeRange from pastEventEnd to currentEventStart is
+   * added to the list of available times.
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:45]: When commenting a function, you don't have to go step-by-step what it does. But you should describe the input and the expected output, and any unexpected behavior / caveats. For example:

Given a set of events for the day and a meeting request, returns a set of free time ranges during which the requested meeting could be scheduled. If no valid time range is found, returns an empty list.

Or something along those lines. If someone is curious about the implementation details, they can look into the function.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 20]
```diff
@@ -14,10 +14,118 @@
...
+   * m: number of available time ranges
+   * O(m+n)
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<Event> eventList = new ArrayList(events);
+    // Here eventList is sorted by event start time in ascending order.
+    Collections.sort(eventList, eventComparator);
+    // Since there are no previous events yet, the start of day is also the
+    // start of the first range to check available time.
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:63]: nit: Sometimes when there are lots of comments, it's good to add an extra line space before the comment so it's clear it's referring to the line after the comment, rather than before.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 21]
```diff
@@ -14,10 +14,118 @@
...
+    int pastEventEnd = TimeRange.START_OF_DAY;
+    int currentEventStart;
+    int availableTimeDuration;
+    List<TimeRange> availableTimes = new ArrayList<>();
+    for (Event event : eventList) {
+      // If the event does not share attendes with the request, it means that
+      // it does not change time availability, so it's skipped.
+      if (!eventSharesAttendeesWithRequest(event, request)) continue;
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:71]: Please use braces - https://google.github.io/styleguide/javaguide.html#s4.1.1-braces-always-used

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 22]
```diff
@@ -14,10 +14,118 @@
...
+      // Available time goes from the end of the previous event to the start of
+      // the current event.
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:73]: I'm glad you added a lot of comments! Some of these comments probably aren't necessary - it's a tough balance :) But in this case, you have later `availableTimeDuration = currentEventStart - pastEventEnd;` - this is clear enough that it doesn't need to be commented (which you were able to do by having good variable names, so good work with that!). Think about for which comments this is true, and which comments are truly clarifying.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 23]
```diff
@@ -14,10 +14,118 @@
...
+      currentEventStart = event.getWhen().start();
+      availableTimeDuration = currentEventStart - pastEventEnd;
+      // This case is if the current event starts after the previous one ends.
+      // Example: |--A--|   |--B--|
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:77]: nit: put comments within the blocks that they are referring. So for instance, this would be:

```
if (currentEventStart > pastEventEnd) {
     // This case is if the current event starts after the previous one ends.
     // Example: |--A--|   |--B--|
      if (availableTimeDuration >= request.getDuration()) {
          availableTimes.add(TimeRange.fromStartDuration(
              pastEventEnd, availableTimeDuration));
      }
      // Regadrdless of whether the duration was enough for the meeting
     // the next available time will start when the current event ends.
     pastEventEnd = event.getWhen().end();
} 
```

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 24]
```diff
@@ -14,10 +14,118 @@
...
+      // we just change the beginning of the next available time.
+      // Example: |--A--|--B--|
+      else if (currentEventStart == pastEventEnd) {
+        pastEventEnd = event.getWhen().end();
+      }
+      // This case is if the current event begins before the previous one ends
+      // and the current event isn't absorbed by the previous one, meaning that
+      // the next beginning of the next available time is changed.
+      // Example: |--A--|
+      //              |--B--|
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:97]: I think these diagrams are helpful, thanks!

**Review by timothy-parsons — CHANGES_REQUESTED** [review]
Take a look at some of these comments, this is already much better, thanks for the changes :)

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 18 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:50]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 19 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:45]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 21 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:71]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 20 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:63]: Got it

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 22 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:73]: Alright, it is a bit hard to find that balance ahaha I hope I keep getting better at it, done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 23 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:77]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 24 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:97]: Yeah, I figured they are a better explanation than words

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 25]
```diff
@@ -14,10 +14,108 @@
...
+    // We pass the events into a list since the collection can't be sorted,
+    // and we sort the list by event start time in ascending order.
+    List<Event> eventList = new ArrayList(events);
+    Collections.sort(eventList, eventComparator);
+
+    // Since there are no previous events yet, the start of day is also the
+    // start of the first range to check available time.
+    int pastEventEnd = TimeRange.START_OF_DAY;
+    int currentEventStart;
+    int availableTimeDuration;
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:62]: I think these are only used in your for-loop, so you can initialize them inside your for loop and keep them as local variables within your for-loop!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 26]
```diff
@@ -14,10 +14,108 @@
...
+      // completely absorbed by the previous event, meaning that the next
+      // available time still starts when the previous event ends.
+      // Example: |----A----|
+      //            |--B--|
+    }
+
+    // Now we just need to check the time from the end of the last meeting
+    // until end of day, basically doing the same procedure as before.
+    currentEventStart = TimeRange.END_OF_DAY;
+    if (currentEventStart - pastEventEnd >= request.getDuration()) {
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:101]: Perhaps just say `if (TimeRange.END_OF_DAY - pastEventEnd >= request.getDuration())...`, then you can keep currentEventStart as a local variable in your for loop.

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 27]
```diff
@@ -14,10 +14,108 @@
...
+      } 
+      else if (currentEventStart == pastEventEnd) {
+        // Example: |--A--|--B--|
+        pastEventEnd = event.getWhen().end();
+      }
+      else if (event.getWhen().end() > pastEventEnd) {
+        // Example: |--A--|
+        //              |--B--|
+        pastEventEnd = event.getWhen().end();
+      }
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:89]: These two else if statements do the same thing, so you can combine them!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java [hunk 28]
```diff
@@ -14,10 +14,108 @@
...
+    List<TimeRange> availableTimes = new ArrayList<>();
+    for (Event event : eventList) {
+      if (!eventSharesAttendeesWithRequest(event, request)) {
+        // If the event does not share attendes with the request, it means that
+        // it does not change time availability, so it's skipped.
+        continue;
+      }
+      currentEventStart = event.getWhen().start();
+      availableTimeDuration = currentEventStart - pastEventEnd;
+
```
Comment (timothy-parsons) [inline walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:72]: nit: probably can remove this empty line. If anything, put an empty line before 'currentEventStart =...', because that line and the following if statements all are directly related logic to each other.

**Review by timothy-parsons — CHANGES_REQUESTED** [review]
Hey Shaar, this looks great! Love the documentation :) I gave you a few small suggestions, but otherwise lgtm!

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 28 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:72]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 27 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:89]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 26 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:101]: Done

File: walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java (hunk 25 above)
Comment (shaargtz) [reply walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java]: Done

**Review by timothy-parsons — APPROVED** [review]
Looks great, thanks for all your changes!
